### PR TITLE
CLN: short-circuit case in Block.replace

### PR DESCRIPTION
**Pull request by jbrockmendel**
Running out of try/excepts to get rid of

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -798,14 +801,27 @@ def replace(
...
+            #   0.5 with int64-dtype
+            #   np.nan with int64-dtype
         except (TypeError, ValueError):
             # GH 22083, TypeError or ValueError occurred within error handling
             # causes infinite loop. Cast and retry only if not objectblock.
             if is_object_dtype(self):
                 raise
 
+            assert not self._can_hold_element(value), value
+
```
Comment (gfyoung) [inline pandas/core/internals/blocks.py]: Tests?  Does this bubble up?

**Comment by jreback** [comment]
any perf improvement? or just code clean

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py]: This is hit in tests.  The original hope was to replace the try/except a few lines above with a `if self._can_hold_element(value):` but as noted in comments a few lines up, the try works in some cases where _can_hold_element fails

**Comment by jbrockmendel** [comment]
> any perf improvement? or just code clean

Only in corner cases

**Comment by gfyoung** [comment]
> Only in corner cases

Worth including in benchmarks?

**Comment by jbrockmendel** [comment]
> Worth including in benchmarks?

I'm not inclined to, no.

**Review by TomAugspurger — APPROVED** [review]
@gfyoung any other concerns?

**Review by gfyoung — APPROVED** [review]
LGTM

**Comment by TomAugspurger** [comment]
Thanks all.

**Pull request merged**